In [172]:
import pandas as pd


In [173]:
df = pd.read_csv('income_train.csv')

In [174]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 143 entries, Id to Target
dtypes: float64(8), int64(130), object(5)
memory usage: 10.4+ MB


In [175]:
df.columns

Index(['Id', 'v2a1', 'hacdor', 'rooms', 'hacapo', 'v14a', 'refrig', 'v18q',
       'v18q1', 'r4h1',
       ...
       'SQBescolari', 'SQBage', 'SQBhogar_total', 'SQBedjefe', 'SQBhogar_nin',
       'SQBovercrowding', 'SQBdependency', 'SQBmeaned', 'agesq', 'Target'],
      dtype='object', length=143)

In [176]:
df


Id      v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  \
0     ID_279628684  190000.0       0      3       0     1       1     0   
1     ID_f29eb3ddd  135000.0       0      4       0     1       1     1   
2     ID_68de51c94       NaN       0      8       0     1       1     0   
3     ID_d671db89c  180000.0       0      5       0     1       1     1   
4     ID_d56d6f5f5  180000.0       0      5       0     1       1     1   
...            ...       ...     ...    ...     ...   ...     ...   ...   
9552  ID_d45ae367d   80000.0       0      6       0     1       1     0   
9553  ID_c94744e07   80000.0       0      6       0     1       1     0   
9554  ID_85fc658f8   80000.0       0      6       0     1       1     0   
9555  ID_ced540c61   80000.0       0      6       0     1       1     0   
9556  ID_a38c64491   80000.0       0      6       0     1       1     0   

      v18q1  r4h1  ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  \
0       NaN     0  ...          100    1849               1        100   
1       1.0     0  ...          144    4489               1        144   
2       NaN     0  ...          121    8464               1          0   
3       1.0     0  ...           81     289              16        121   
4       1.0     0  ...          121    1369              16        121   
...     ...   ...  ...          ...     ...             ...        ...   
9552    NaN     0  ...           81    2116              25         81   
9553    NaN     0  ...            0       4              25         81   
9554    NaN     0  ...           25    2500              25         81   
9555    NaN     0  ...          121     676              25         81   
9556    NaN     0  ...           64     441              25         81   

      SQBhogar_nin  SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
0                0         1.000000         0.0000   100.0000   1849       4  
1                0         1.000000        64.0000   144.0000   4489       4  
2                0         0.250000        64.0000   121.0000   8464       4  
3                4         1.777778         1.0000   121.0000    289       4  
4                4         1.777778         1.0000   121.0000   1369       4  
...            ...              ...            ...        ...    ...     ...  
9552             1         1.562500         0.0625    68.0625   2116       2  
9553             1         1.562500         0.0625    68.0625      4       2  
9554             1         1.562500         0.0625    68.0625   2500       2  
9555             1         1.562500         0.0625    68.0625    676       2  
9556             1         1.562500         0.0625    68.0625    441       2  

[9557 rows x 143 columns]

In [177]:
df.describe(include= 'object')

Id    idhogar dependency edjefe edjefa
count           9557       9557       9557   9557   9557
unique          9557       2988         31     22     22
top     ID_edc3d98de  fd8a6d014        yes     no     no
freq               1         13       2192   3762   6230

# Checking for biases in dataset

In [178]:
df.Target.value_counts(normalize=True)

4    0.627394
2    0.167103
3    0.126504
1    0.079000
Name: Target, dtype: float64

# There IS bias in the data set since target column is not equally distributed

# Checking family id number column to check if all members in a family have same poverty level

In [184]:
df.idhogar.value_counts()

fd8a6d014    13
0c7436de6    12
ae6cf0558    12
3fe29a56b    11
4476ccd4c    11
             ..
84705aa3b     1
7fc5dc154     1
53bab505c     1
72160432f     1
e79405ae7     1
Name: idhogar, Length: 2988, dtype: int64

In [185]:
df_fam= df.groupby(['idhogar', 'Target']).size().reset_index(name= 'counts')
df_fam

idhogar  Target  counts
0     001ff74ca       4       2
1     003123ec2       2       4
2     004616164       2       2
3     004983866       3       2
4     005905417       2       3
...         ...     ...     ...
3069  ff9343a35       4       4
3070  ff9d5ab17       4       3
3071  ffae4a097       4       2
3072  ffe90d46f       1       4
3073  fff7d6be1       4       4

[3074 rows x 3 columns]

In [186]:
df_fam.idhogar.value_counts()

5c6f32bbc    3
3c73c107f    2
564eab113    2
6833ac5dc    2
811a35744    2
            ..
2811494b7    1
c0480739d    1
dc83157df    1
f13d8a988    1
e79405ae7    1
Name: idhogar, Length: 2988, dtype: int64

In [187]:
df_fam_dup= df_fam[df_fam.duplicated(['idhogar'])]

In [188]:
df_fam_dup.head(86)

idhogar  Target  counts
18    0172ab1d9       3       3
44    03f4e5f4d       2       1
57    0511912b6       4       3
96    078a0b6e2       2       2
127   09e25d616       2       3
...         ...     ...     ...
2748  e65d4b943       3       4
2858  efd3aec61       3       2
2864  f006348ed       3       2
2951  f7b421c2c       2       4
2973  f94589d38       2       2

[86 rows x 3 columns]

In [189]:
df_fam_dup1= df_fam_dup.drop_duplicates(['idhogar'])

In [190]:
fam_id = df_fam_dup1.idhogar.tolist()

In [191]:
#fam_id

In [192]:
len(fam_id)

85

# The following household is an example of a household with multiple family members not having the same poverty level

In [193]:
filtered_data = df[df['idhogar']== '5c6f32bbc' ]

In [194]:
filtered_data= filtered_data[['idhogar', 'Target']]

In [195]:
filtered_data


idhogar  Target
6023  5c6f32bbc       3
6024  5c6f32bbc       4
6025  5c6f32bbc       2

In [196]:
df1 = df.copy() #copying the original dataframe for setting the target value of each household member to same level

In [198]:
df1.iloc[1]['idhogar'] #checking value of household_id first row

'0e5d7a658'

In [199]:
df1.idhogar.isnull().sum() #making sure there are no null values in column

0

# Replacing the poverty level of each member in a household to the same value

In [200]:
%%time
def fix_pov_level(ID):
    for i in range (0, 9556):
        if df1.iloc[i]['idhogar'] == ID:
            target_val = df1.loc[df1.idhogar == ID, 'Target'].tolist()[0]
            index_list= df1.loc[df1.idhogar == ID].index.values.tolist()
            #print('id is ' +str(ID))
            for index in index_list:
                df1.at[index,'Target'] = target_val
        i+=1 
        
      
    return  

for counter in range(0, len(fam_id)-1):
    fix_pov_level(fam_id[counter])
    counter +=1
    


Wall time: 3min


In [208]:
df1[df1['idhogar']== '0172ab1d9'] #checking to see if Target values are set to the same number

Id  v2a1  hacdor  rooms  hacapo  v14a  refrig  v18q  v18q1  \
7651  ID_a5e2b0639   NaN       0      5       0     1       1     0    NaN   
7652  ID_5dfdf4ebe   NaN       0      5       0     1       1     0    NaN   
7653  ID_762e1fd96   NaN       0      5       0     1       1     0    NaN   
7654  ID_c76b87004   NaN       0      5       0     1       1     0    NaN   
7655  ID_1dffe3dcf   NaN       0      5       0     1       1     0    NaN   

      r4h1  ...  SQBescolari  SQBage  SQBhogar_total  SQBedjefe  SQBhogar_nin  \
7651     0  ...           49     196              25         36             4   
7652     0  ...          100     289              25         36             4   
7653     0  ...           36    2601              25         36             4   
7654     0  ...           36    2304              25         36             4   
7655     0  ...          121     441              25         36             4   

      SQBovercrowding  SQBdependency  SQBmeaned  agesq  Target  
7651         2.777778       0.444444  58.777775    196       3  
7652         2.777778       0.444444  58.777775    289       3  
7653         2.777778       0.444444  58.777775   2601       3  
7654         2.777778       0.444444  58.777775   2304       3  
7655         2.777778       0.444444  58.777775    441       3  

[5 rows x 143 columns]

In [211]:
df.parentesco1.value_counts()

0    6584
1    2973
Name: parentesco1, dtype: int64

# Finding if df_fam has a missing head of family i.e parentesco1 is not 1 in any of the family members

# Creating new dataframe grouping by household id and parentesco1 i.e. is a household head

In [212]:
df_fam1 = df.groupby(['idhogar', 'parentesco1']).size().reset_index(name= 'counts')

In [213]:
df_fam1

idhogar  parentesco1  counts
0     001ff74ca            0       1
1     001ff74ca            1       1
2     003123ec2            0       3
3     003123ec2            1       1
4     004616164            0       1
...         ...          ...     ...
5568  ffae4a097            1       1
5569  ffe90d46f            0       3
5570  ffe90d46f            1       1
5571  fff7d6be1            0       3
5572  fff7d6be1            1       1

[5573 rows x 3 columns]

In [214]:
#df_fam1.counts.sum() 
#checking the total number of values with original dataframe

In [250]:
fam_id_list = df_fam1['idhogar'].tolist() #creating a list of family ids extracted from from the column idhogar


In [216]:
len(fam_id_list)

5573

In [217]:
head_ls = []  
not_head_ls = []
counter = 1
current_fam = fam_id_list[0]
not_head_ls.append(fam_id_list[0])
for value in range (1, len(fam_id_list)):
    new_fam = fam_id_list[value]
    
    if new_fam == current_fam:
        head_ls.append(fam_id_list[value])
    else: not_head_ls.append(fam_id_list[value])
    counter +=1
    current_fam = new_fam    


In [218]:
len(head_ls)

2585

In [219]:
len(not_head_ls)

2988

# The list not_head_ls contains the grouped households that are not a head 
# The list head_ls contains households that is a head
# Since the list not_head_ls contains more number of households (2988) than the list head_ls (2585), there are 403 households with no head

In [220]:
#Checking missing values in columns
df.isna().sum()

Id                    0
v2a1               6860
hacdor                0
rooms                 0
hacapo                0
                   ... 
SQBovercrowding       0
SQBdependency         0
SQBmeaned             5
agesq                 0
Target                0
Length: 143, dtype: int64

# counting the null values in the dataframe df1

In [221]:
df1.isnull().sum()

Id                    0
v2a1               6860
hacdor                0
rooms                 0
hacapo                0
                   ... 
SQBovercrowding       0
SQBdependency         0
SQBmeaned             5
agesq                 0
Target                0
Length: 143, dtype: int64

In [222]:
df1.Target.isnull().sum() #no  null values in Target column

0

In [223]:
test_df = pd.read_csv('income_test.csv')

# Preparing df for RF 

In [224]:
df1.drop(['Id', 'idhogar'], axis =1, inplace= True)


In [225]:
df1.edjefe.unique()

array(['10', '12', 'no', '11', '9', '15', '4', '6', '8', '17', '7', '16',
       '14', '5', '21', '2', '19', 'yes', '3', '18', '13', '20'],
      dtype=object)

In [226]:
df1.edjefe.replace(to_replace=['no', 'yes'], value=['0', '1'], inplace= True)



In [227]:
df1['edjefe']= df1['edjefe'].astype(str).astype(int)

In [228]:
df1.edjefa.replace(to_replace=['no', 'yes'], value=['0', '1'], inplace= True)
df1['edjefa']= df1['edjefa'].astype(str).astype(int)

In [229]:
df1.dependency.replace(to_replace=['no', 'yes'], value=['0', '1'], inplace= True)
df1['dependency']= df1['dependency'].astype(str).astype(float)

In [230]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9557 entries, 0 to 9556
Columns: 141 entries, v2a1 to Target
dtypes: float64(9), int32(2), int64(130)
memory usage: 10.2 MB


In [231]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23856 entries, 0 to 23855
Columns: 142 entries, Id to agesq
dtypes: float64(8), int64(129), object(5)
memory usage: 25.8+ MB


In [232]:
df1.drop(['v2a1', 'rez_esc', ], axis=1, inplace= True)

In [233]:
df1.isnull().sum().sum()

7352

In [234]:
df1.drop(['v18q1'], axis =1, inplace=True)

In [235]:
df1.isna().sum().sum()

10

In [236]:
df1= df1.dropna()

In [237]:
df1.isna().sum().sum()

0

# Random Forest Classifier and Grid Search 

In [238]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn import metrics

from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score, roc_curve, classification_report, recall_score, f1_score

In [239]:
used_features = list(df1.columns.values.tolist())
X= df1[used_features].values
y= df1['Target']

In [240]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 1)

In [241]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(7641, 138)
(7641,)
(1911, 138)
(1911,)


In [242]:
classifier.fit(X_train, y_train)
y_predict = classifier.predict(X_test)
accuracy_rf = accuracy_score(y_predict, y_test)
print(accuracy_rf)


0.9937205651491365


In [243]:
param_grid = {'n_estimators': [100 ,150,250,350],
              'criterion': ['gini' , 'entropy'] ,
              'max_depth' : [3, 5 , 7,10] ,
              'min_samples_split' : [20 ,40 , 60]
             } 

In [244]:
grid_rf = GridSearchCV( classifier ,param_grid, refit = True, verbose = 1 , n_jobs = -1 , cv = 2)
grid_rf.fit(X_train,y_train)

Fitting 2 folds for each of 96 candidates, totalling 192 fits


GridSearchCV(cv=2, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 5, 7, 10],
                         'min_samples_split': [20, 40, 60],
                         'n_estimators': [100, 150, 250, 350]},
             verbose=1)

In [245]:
grid_predictions = grid_rf.predict(X_test)

In [246]:
print(confusion_matrix(y_test,grid_predictions))

[[  95   47    0    8]
 [   0  288    0   16]
 [   0    0  224    5]
 [   0    0    0 1228]]


In [247]:
print(classification_report(y_test,grid_predictions))

              precision    recall  f1-score   support

           1       1.00      0.63      0.78       150
           2       0.86      0.95      0.90       304
           3       1.00      0.98      0.99       229
           4       0.98      1.00      0.99      1228

    accuracy                           0.96      1911
   macro avg       0.96      0.89      0.91      1911
weighted avg       0.96      0.96      0.96      1911



In [248]:
grid_rf.best_params_

{'criterion': 'entropy',
 'max_depth': 10,
 'min_samples_split': 20,
 'n_estimators': 350}

In [249]:
classifier_best = RandomForestClassifier(criterion='entropy',max_depth=10,min_samples_split= 20,n_estimators= 150)
classifier_best.fit(X_train, y_train)
y_predict = classifier_best.predict(X_test)
accuracy_rf_best = accuracy_score(y_predict, y_test)
print(accuracy_rf_best)


0.9555206698063841
